[iOS] Text selection highlight does not honor layer occlusion/clipping

whsieh opened the pull request:
#### b7fc2d972516a068452fd5676e26e8a354841670
<pre>
[iOS] Text selection highlight does not honor layer occlusion/clipping
<a href="https://bugs.webkit.org/show_bug.cgi?id=281601">https://bugs.webkit.org/show_bug.cgi?id=281601</a>
<a href="https://rdar.apple.com/107058373">rdar://107058373</a>

Reviewed by Abrar Rahman Protyasha.

Currently, the &quot;Selection Honors Overflow Scrolling&quot; flag allows us to slot native selection views
underneath composited subscrollable areas (i.e. `WKChildScrollView`s), by plumbing the enclosing
scrolling node ID over to the UI process via `EditorState` with each layer tree transaction.

While this enables selection UI in subscrollable containers to scroll smoothly along with the
selected content and also fixes occlusion in the case where occluding content is in a compositing
view that covers the selection in a subscrollable container, this approach *doesn&apos;t* address the
more general problem of native selection UI appearing over content that should otherwise occlude the
selection, if:

1.  The selection is in the main frame, or
2.  The content that would otherwise occlude the selection is in the same scrollable container as
    the selection.

To fix *most* (not all) of these remaining scenarios, we make further changes to the &quot;Selection
Honors Overflow Scrolling&quot; behavior, such that we no longer propagate a scrolling node ID via
`EditorState` to the UI process; instead, we now propagate the ID of the graphics layer that backs
the selection&apos;s `RenderLayer` (or the scrolled contents layer, if appropriate), and then map this ID
to a platform compositing view which we use as the parent for the selection views.

See below for more details.

* LayoutTests/editing/selection/ios/selection-highlight-in-fixed-position-container-expected.txt: Added.
* LayoutTests/editing/selection/ios/selection-highlight-in-fixed-position-container.html: Added.

Add a layout test to verify that a selection inside a fixed-position container stays on screen when
scrolling.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async longPressElement):

Make this work when long pressing content inside fixed-position containers, by using
`getBoundingClientRect()`.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Expose this on `WebCore::Settings` too, so that we don&apos;t need to store a flag on `WebPage` when
checking this state.

* Source/WebCore/dom/Position.h:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderObject.h:

Shuffle around some `WEBCORE_EXPORT`s, some of which are no longer needed.

* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Replace the `ScrollingNodeID` with a `PlatformLayerIdentifier`.

* Source/WebKit/UIProcess/ios/UIKitUtilities.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(-[UIView _wk_parentScrollView]):
(-[UIView _wk_isAncestorOf:]): Deleted.

Replace this category helper method with a new one, which finds the ancestor scroller of a given
view.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView scrolledContentView]): Deleted.

Remove this now-unused helper method.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldHideSelectionWhenScrolling]):

Now that we (more) correctly paint native selection UI in fixed-position containers, we can stop
hiding the selection every time we scroll now, when the selection is inside fixed-position content.

(-[WKContentView _pointIsInsideSelectionRect:outBoundingRect:]):

Make a small adjustment here to ensure that hit-testing for the purposes of toggling the edit menu
still honors overflow scrolling clip.

(-[WKContentView _selectionContainerViewInternal]):

Make this grab the corresponding layer tree node for the platform layer ID that encloses the
selection endpoints, and use that as the selection container.

(-[WKContentView _selectionContainerScrollView]): Deleted.
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper prepareToMoveSelectionContainer:]):

Additionally make sure that the selection views are hosted above the tile grid controller container
layer; this is necessary to ensure that selection views show up above selected content, but under
any composited content (fixed/sticky-position banners, z-ordered elements, etc.) that should occlude
it.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Add a helper method to search up the layer tree for the closest ancestor backed by a graphics layer.

(WebKit::enclosingScroller): Deleted.
(WebKit::forEachEnclosingScroller): Deleted.
(WebKit::WebPage::computeSelectionClipRectAndEnclosingScroller const): Deleted.

Canonical link: <a href="https://commits.webkit.org/285350@main">https://commits.webkit.org/285350@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c824b8c6ddefeb1e8813721de739c68515929b28

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25123 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76510 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23544 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59560 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23366 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15508 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46875 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62297 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43528 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19768 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21894 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65464 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20126 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78182 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71589 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16576 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62315 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15981 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12969 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6602 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93370 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47554 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20546 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48623 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49911 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48366 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/editing/selection/ios/selection-highlight-in-fixed-position-container.html:
```diff
@@ -0,0 +1,58 @@
+    width: 100%;
+    height: 100px;
+    top: 0;
+    background: tomato;
```
TIL

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -5558,99 +5557,57 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
 
-        if (innerScrollingClipRect.isEmpty())
-            innerScrollingClipRect = clipRectInRootView;
+    for (; enclosingLayer; enclosingLayer = enclosingLayer->enclosingContainingBlockLayer(CrossFrameBoundaries::Yes)) {
```
I think you can fold this together

```
for (CheckedPtr enclosingLayer = startEnclosingLayer->commonAncestorWithLayer(*endEnclosingLayer); enclosingLayer; enclosingLayer = enclosingLayer->enclosingContainingBlockLayer(CrossFrameBoundaries::Yes)) {
    ...
}
```

aprotyas replied:
(if you think this doesn't hurt readability)

whsieh replied:
Done!

whsieh committed:
[iOS] Text selection highlight does not honor layer occlusion/clipping
https://bugs.webkit.org/show_bug.cgi?id=281601
rdar://107058373

Reviewed by Abrar Rahman Protyasha.

Currently, the "Selection Honors Overflow Scrolling" flag allows us to slot native selection views
underneath composited subscrollable areas (i.e. `WKChildScrollView`s), by plumbing the enclosing
scrolling node ID over to the UI process via `EditorState` with each layer tree transaction.

While this enables selection UI in subscrollable containers to scroll smoothly along with the
selected content and also fixes occlusion in the case where occluding content is in a compositing
view that covers the selection in a subscrollable container, this approach *doesn't* address the
more general problem of native selection UI appearing over content that should otherwise occlude the
selection, if:

1.  The selection is in the main frame, or
2.  The content that would otherwise occlude the selection is in the same scrollable container as
    the selection.

To fix *most* (not all) of these remaining scenarios, we make further changes to the "Selection
Honors Overflow Scrolling" behavior, such that we no longer propagate a scrolling node ID via
`EditorState` to the UI process; instead, we now propagate the ID of the graphics layer that backs
the selection's `RenderLayer` (or the scrolled contents layer, if appropriate), and then map this ID
to a platform compositing view which we use as the parent for the selection views.

See below for more details.

* LayoutTests/editing/selection/ios/selection-highlight-in-fixed-position-container-expected.txt: Added.
* LayoutTests/editing/selection/ios/selection-highlight-in-fixed-position-container.html: Added.

Add a layout test to verify that a selection inside a fixed-position container stays on screen when
scrolling.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async longPressElement):

Make this work when long pressing content inside fixed-position containers, by using
`getBoundingClientRect()`.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Expose this on `WebCore::Settings` too, so that we don't need to store a flag on `WebPage` when
checking this state.

* Source/WebCore/dom/Position.h:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderObject.h:

Shuffle around some `WEBCORE_EXPORT`s, some of which are no longer needed.

* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Replace the `ScrollingNodeID` with a `PlatformLayerIdentifier`.

* Source/WebKit/UIProcess/ios/UIKitUtilities.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(-[UIView _wk_parentScrollView]):
(-[UIView _wk_isAncestorOf:]): Deleted.

Replace this category helper method with a new one, which finds the ancestor scroller of a given
view.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView scrolledContentView]): Deleted.

Remove this now-unused helper method.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldHideSelectionWhenScrolling]):

Now that we (more) correctly paint native selection UI in fixed-position containers, we can stop
hiding the selection every time we scroll now, when the selection is inside fixed-position content.

(-[WKContentView _pointIsInsideSelectionRect:outBoundingRect:]):

Make a small adjustment here to ensure that hit-testing for the purposes of toggling the edit menu
still honors overflow scrolling clip.

(-[WKContentView _selectionContainerViewInternal]):

Make this grab the corresponding layer tree node for the platform layer ID that encloses the
selection endpoints, and use that as the selection container.

(-[WKContentView _selectionContainerScrollView]): Deleted.
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper prepareToMoveSelectionContainer:]):

Additionally make sure that the selection views are hosted above the tile grid controller container
layer; this is necessary to ensure that selection views show up above selected content, but under
any composited content (fixed/sticky-position banners, z-ordered elements, etc.) that should occlude
it.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Add a helper method to search up the layer tree for the closest ancestor backed by a graphics layer.

(WebKit::enclosingScroller): Deleted.
(WebKit::forEachEnclosingScroller): Deleted.
(WebKit::WebPage::computeSelectionClipRectAndEnclosingScroller const): Deleted.

Canonical link: https://commits.webkit.org/285350@main
